src/communication/RpcClient.cpp

```diff
@@ -10,3 +10,371 @@
+	std::function<void(communication::RpcClient::Status)> expire;
+	size_t instance_id;
+
+	auto operator>(const PendingRequest& other) const;
+};
+
+struct ScrubablePendingQueue
+    : public std::priority_queue<PendingRequest, std::vector<PendingRequest>,
+                                 std::greater<PendingRequest>> {
+	~ScrubablePendingQueue();
+	auto scrub(size_t instance_id);
+	PendingRequest& top();
+};
+
+struct ExpireWorker {
+	ExpireWorker();
+	~ExpireWorker();
+	void enqueue(PendingRequest&& request);
+	void scrub(size_t instance_id);
+	void doWork();
+
+private:
+	std::mutex pending_mtx_;
+	ScrubablePendingQueue pending_;
+	std::thread worker_;
+	std::atomic<bool> stop_{false};
+	std::condition_variable wake_worker_;
+};
+
+}  // namespace detail
+}  // namespace
+
+namespace uprotocol::communication {
+
+////////////////////////////////////////////////////////////////////////////////
+struct RpcClient::ExpireService {
+	ExpireService() : instance_id_(next_instance_id++) {}
+
+	~ExpireService() { worker.scrub(instance_id_); }
+
+	void enqueue(std::chrono::steady_clock::time_point when_expire,
+	             transport::UTransport::ListenHandle&& response_listener,
+	             std::function<void(RpcClient::Status)> expire) {
+		detail::PendingRequest pending{
+		    .when_expire = when_expire,
+		    .response_listener = std::move(response_listener),
+		    .expire = std::move(expire),
+		    .instance_id = instance_id_};
+
+		worker.enqueue(std::move(pending));
+	}
+
+private:
+	static inline std::atomic<size_t> next_instance_id{0};
+	static inline detail::ExpireWorker worker;
+	size_t instance_id_;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+RpcClient::RpcClient(std::shared_ptr<transport::UTransport> transport,
+                     v1::UUri&& method, v1::UPriority priority,
+                     std::chrono::milliseconds ttl,
+                     std::optional<v1::UPayloadFormat> payload_format,
+                     std::optional<uint32_t> permission_level,
+                     std::optional<std::string> token)
+    : transport_(transport),
+      ttl_(ttl),
+      builder_(datamodel::builder::UMessageBuilder::request(
+          std::move(method), v1::UUri(transport_->getDefaultSource()), priority,
+          ttl_)),
+      expire_service_(std::make_unique<ExpireService>()) {
+	if (payload_format) {
+		builder_.withPayloadFormat(*payload_format);
+	}
+
+	if (permission_level) {
+		builder_.withPermissionLevel(*permission_level);
+	}
+
+	if (token) {
+		builder_.withToken(*token);
+	}
+}
+
+RpcClient::InvokeHandle RpcClient::invokeMethod(v1::UMessage&& request,
+                                                Callback&& callback) {
+	auto when_expire = std::chrono::steady_clock::now() + ttl_;
+	auto reqid = request.attributes().id();
+	// Used to ensure that, no matter what, the callback is only called once
+	auto callback_once = std::make_shared<std::once_flag>();
```

Comment:
Please improve this comment to include the fact that there is a race condition between the "expire logic" and the "received response" logic.  As it stands it almost implies that you don't know (or aren't sure) why some weird race condition exists.

Reply:
Fixed with next update to PR